hw/ip/otbn/dv/rig/rig/gens/branch_gen.py

```diff
@@ -0,0 +1,133 @@
+        self.bne = self._get_named_insn(insns_file, 'bne')
+
+        # beq and bne expect operands: grs1, grs2, offset
+        for insn in [self.beq, self.bne]:
+            if not (len(insn.operands) == 3 and
+                    isinstance(insn.operands[0].op_type, RegOperandType) and
+                    insn.operands[0].op_type.reg_type == 'gpr' and
+                    not insn.operands[0].op_type.is_dest() and
+                    isinstance(insn.operands[1].op_type, RegOperandType) and
+                    insn.operands[1].op_type.reg_type == 'gpr' and
+                    not insn.operands[1].op_type.is_dest() and
+                    isinstance(insn.operands[2].op_type, ImmOperandType) and
+                    insn.operands[2].op_type.signed):
+                raise RuntimeError('{} instruction from instructions file is not '
+                                   'the shape expected by the Branch generator.'
+                                   .format(insn.mnemonic))
+
+        self.imm_op_type = self.bne.operands[2].op_type
+
+        for insn in insns_file.insns:
+            if insn.mnemonic in ['beq', 'bne']:
+                weight = cfg.insn_weights.get(insn.mnemonic)
+                if weight > 0:
+                    self.weights.append(weight)
+                    self.insns.append(insn)
+
+        # Disable this generator if no branch instruction has a positive weigh
+        assert len(self.insns) == len(self.weights)
+        if not self.weights:
+            self.disabled = True
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+
+        # Pick maximum and minimum address range for the current PC.
+        imm_rng = self.imm_op_type.get_op_val_range(model.pc)
+        assert imm_rng is not None
+
+        min_addr, max_addr = imm_rng
+
+        # Get known registers. We always have x0.
+        # So it should never fail.
+        known_regs = model.regs_with_known_vals('gpr')
+        assert known_regs
+
+        # Pick a random register among known registers.
+        idx, value = random.choice(known_regs)
+
+        # Get the chosen base register index as grs1 and grs2 operand. This is
+        # because we want to branch to the faulty addresses with this snippet.
+        op_val_grs1 = idx
+        assert op_val_grs1 is not None
+
+        equals = []
+        not_equals = []
+
+        for reg_idx, reg_val in known_regs:
+            if reg_val == value:
+                equals.append(reg_idx)
+            else:
+                not_equals.append(reg_idx)
+        # Pick the offset address
+        off_enc = self.cal_offset(min_addr, max_addr, program, model)
+        if off_enc is None:
+            return None
+
+        # Pick the instruction and second operand
+        op_val_grs2, chosen_insn = self.cal_second_op(equals, not_equals)
+        assert op_val_grs2 is not None
+        assert chosen_insn is not None
```

Comment:
This function should have return type something like `Optional[Tuple[int, Insn]]` and it *can* return `None`. So I'd do something like:
```
    ret = self.cal_second_op(equals, not_equals)
    if ret is None:
        return None
    op_val_grs2, chosen_insn = ret
```
(which will also stop MyPy moaning about things :-))